src/components/servo/client.hpp

```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include <grpc/support/log.h>
+#include <grpcpp/grpcpp.h>
+
+#include "servo.hpp"
+#include "component/servo/v1/servo.grpc.pb.h"
+
+using grpc::Channel;
+using grpc::ClientContext;
+using viam::component::servo::v1::ServoService;
+
+class ServoClient: public Servo {
+    public:
+        explicit ServoClient(std::string name_, std::shared_ptr<Channel> channel);
+        static std::shared_ptr<ServoClient> from_robot(std::shared_ptr<RobotClient> robot, std::string name);
+        void move(int angle);
```

Comment:
These are degrees. I don't know anything about `boost::unit` but a quick look makes me think we could, but would it hurt ergonomics? I don't know how difficult it is to pass in a `boost::units::degree` vs a simple `int`

Reply:
I think that long-term using something like `boost::unit` would be a *very* good idea, but I don't think this review is the place. I've noted it on my recommendations doc so we don't forget.

Meanwhile, I don't know what degree of flexibility there is around the names of these methods. I can imagine that there could be rules about naming consistency across different language SDKs.

But if you do have flexibility around naming, and in the absence of `boost::unit`, I'd recommend folding the intended type into either the method name or the parameter. Something like:

```
void move(int degrees);
```

Or

```
void move_degrees(int angle)
```

Also, is `int` the right type here? Should it be `double`? Presumably it makes sense to move `22.5` degrees.

Similarly `get_position` should probably describe the units it returns?

```
int get_position_degrees();  
```

Etc.

Reply:
I really like it as the parameter name, not least because we're trying not to have too much deviation between the proto method (`Move`) and the SDK methods. 

Which makes `get_position` harder, since the proto name is simply `GetPosition`. The difference is that in proto, the return type is an object with the attribute `position_deg` (or `angle_deg` for `Move`) which accurately describes the unit, whereas in the various SDKs, it returns the value directly (not wrapped in an object). So we lose that descriptive ability. 

Right now we rely on documentation to convey units. So I think this is a larger discussion of how we want to handle units. Like perhaps we use type aliases in languages that support them? 

Also it should not be `double`, but perhaps it shouldn't be `int` either since the proto uses `uint32`

Reply:
> I really like it as the parameter name, not least because we're trying not to have too much deviation between the proto method (`Move`) and the SDK methods.

Can you elaborate on why you want to keep them so close? My understanding is that the proto layer is the "behind the scenes" implementation, so I don't see why, necessarily, the names used there should inform the names/types used in the developer facing SDK.

If it *is* important that the SDK offer names that have a strong mapping to the proto layer, or are more or less a passthrough directly to the proto layer, that doesn't preclude offering the functionality under both the proto-y version and a more sophisticated version:

```
void move(uint32 angle);
void move(boost::units::degrees angle);
```

Where the second implementation ultimately delegates to the first.

> 
> Which makes `get_position` harder, since the proto name is simply `GetPosition`. The difference is that in proto, the return type is an object with the attribute `position_deg` (or `angle_deg` for `Move`) which accurately describes the unit, whereas in the various SDKs, it returns the value directly (not wrapped in an object). So we lose that descriptive ability.
> 
> Right now we rely on documentation to convey units. So I think this is a larger discussion of how we want to handle units. Like perhaps we use type aliases in languages that support them?

Yes, and that's basically what `boost::unit` gives you in C++: it describes a bunch of type aliases for templates that enforce dimensional consistency. To be fair, I've never used it, but this sounds like it might be a great use case for it.



> 
> Also it should not be `double`, but perhaps it shouldn't be `int` either since the proto uses `uint32`

That's interesting. Is it scaled somehow? Or is one degree literally `uint32(1)` and 360 degrees is `uint32(360)`? Do servos generally not offer sub-degree precision?

Reply:
The reasoning for keeping names close is mostly for easier understanding/portability between languages. The name in raw proto ≅ name in typescript ≅ name in python ≅ name in golang. That being said, it was only important that the **name** of the function be mapped appropriately. The parameters, on the other hand, can and should be more native. So I think we don't need 2 versions of `move`, but only the more sophisticated `move(boost::...)` version. 

Said another way, it is important that only the name of functions be similar for so that users can find/map them across languages, and the implementation of those functions be as native as possible. 

I like the idea of using type aliases across languages (boost here).

Reply:
Just asked @randhid on scaling degrees -- we only do integer degrees, no sub degrees nor scaling

Reply:
OK, let's just rename the parameter for `move` to `int degrees`. Not much we can do about `position`, for now.